### ISSUE TITLE
wrap_url_and_doi mishandles DOIs in URLs

### ISSUE THREAD

**Issue by dstillman** [Juris-M/citeproc-js#57]
When a style shows a DOI as part of a URL, the embedded DOI gets wrapped instead of the outer URL.

```
<div class="csl-entry">Mueen Ahmed, K., &amp; Al Dhubaib, B. (2011). Zotero: A bibliographic assistant to researcher. <i>Journal of Pharmacology and Pharmacotherapeutics</i>, <i>2</i>(4), 303. https://doi.org/<a href="https://doi.org/10.4103/0976-500X.85940">10.4103/0976-500X.85940</a></div>
```

**Comment by fbennett** [comment]
Oh. That's not good.

**Comment by fbennett** [comment]
Sorry for the delay in getting back. I'm not yet able to reproduce this in the processor. Is the CSL code supplying the protocol stub as a literal prefix?

**Comment by dstillman** [comment]
Ah, sorry — you're right:

https://github.com/citation-style-language/styles/blob/20f642383c16b1f98acc64bc4f6d2cc8b3c5218c/chicago-author-date.csl#L179

Maybe worth special-casing this? It probably doesn't make sense to wrap `doi:` but if the prefix begins with `http` it seems reasonable.

**Comment by fbennett** [comment]
We definitely want to adjust the output there, but I'm not sure exactly how best to handle it. I've posted to the CSL list for thoughts, and when the returns are in we'll get this fixed. (The list now tells me I'm no longer a member, so it might take a few days to clear the moderation barrier.)

**Comment by fbennett** [comment]
@dstillman No action on moderation yet, but I shouldn't use that for an excuse. Tagged release 1.1.182 will wrap the full URL, with some special footwork:
* If the style-provided prefix is `https://doi.org/` a full URL will be wrapped.
* With any other style-provided prefix, only the DOI field content will be wrapped.
* When the special prefix is used in the style:
  - If the field value starts with `http://doi.org/`, it will be normalized to `https://doi.org/`
  - If the field value starts with some other `http://` or `https://` protocol declaration, it will be wrapped as-is

I've added a local (citeproc-js-only) [test fixture](https://github.com/Juris-M/citeproc-js/blob/master/tests/fixtures/local/magic_WrapDOI.txt) that covers most of the conditions above.

I *think* that will do it. If I've missed anything, let me know!

**Comment by dstillman** [comment]
It looks like @adam3smith responded on the list, but this all sounds good. We'll try it out. Thanks!

**Comment by rmzelle** [comment]
I responded to @fbennett's questions on the mailing list at http://xbiblio-devel.2463403.n2.nabble.com/Handling-DOIs-in-a-processor-tp7579530.html.

I checked all the DOI prefixes used in the repository and their frequency. Some things of note:

* In addition to "https://doi.org/", we also have URL prefixes with additional labels or punctuation in the prefix, such as "DOI: https://doi.org/", " https://doi.org/", and "[https://doi.org/" (see below). So I would change @fbennett's rule "If the style-provided prefix is https://doi.org/ a full URL will be wrapped." to "If the style-provided prefix ends on https://doi.org/, a full URL will be wrapped (this URL would exclude the part of the prefix before 'https://doi.org/')."
* I think any type of DOI prefix should always be stripped from field values. "https://doi.org/10.4103/0976-500X.85940", "http://dx.doi.org/10.4103/0976-500X.85940", or "doi:10.4103/0976-500X.85940" should all be stripped to "10.4103/0976-500X.85940".

```
$ egrep -h -i -o '\prefix=".*?doi.*?"' *.csl | sort | uniq -c | sort -bgr
 191 prefix="doi:"
  92 prefix="https://doi.org/"
  68 prefix="doi: "
  52 prefix="Doi: "
  43 prefix="DOI: "
  25 prefix=" doi:"
  15 prefix=", doi:"
  15 prefix=", doi: "
  11 prefix=". doi:"
  10 prefix="DOI "
  10 prefix=" doi: "
   9 prefix=", DOI: "
   9 prefix=" DOI: "
   8 prefix=". doi: "
   7 prefix="DOI&#160;"
   7 prefix=". DOI: "
   6 prefix="DOI: https://doi.org/"
   5 prefix=", DOI:"
   5 prefix="       |doi="
   4 prefix="DOI:"
   4 prefix=" https://doi.org/"
   4 prefix=" &lt;https://doi.org/"
   3 prefix="doi:/"
   2 prefix="doi: https://doi.org/"
   2 prefix="doi "
   2 prefix="DOI-"
   2 prefix=". https://doi.org/"
   2 prefix="(doi:"
   2 prefix="&lt;https://doi.org/"
   2 prefix=" DOI:"
   2 prefix=" DOI: https://doi.org/"
   2 prefix=" (doi:"
   2 prefix=" &lt;doi:"
   1 prefix="https://doi.org"
   1 prefix="doi.org/"
   1 prefix="[https://doi.org/"
   1 prefix="[doi: "
   1 prefix="DOI:https://doi.org/"
   1 prefix="; https://doi.org/"
   1 prefix="; doi:"
   1 prefix="; DOI: "
   1 prefix=": doi:"
   1 prefix=". DOI:https://doi.org/"
   1 prefix="- doi: "
   1 prefix=", https://doi.org/"
   1 prefix=", doi: https://doi.org/"
   1 prefix="(DOI&#160;: "
   1 prefix="&lt;doi:"
   1 prefix="&lt;a href=&quot;https://doi.org/"
   1 prefix="&amp;#10;https://doi.org/"
   1 prefix=" doi : "
   1 prefix=" [https://doi.org/"
   1 prefix=" [doi:"
   1 prefix=" DOI={"
   1 prefix=" DOI:https://doi.org/"
   1 prefix=" DOI "
   1 prefix=" Available at https://doi.org/"
```

**Comment by fbennett** [comment]
Thanks for the nabble link. I've apparently been bumped from the list, so
haven't seen any of the mail traffic.

The list of prefixes shows we should loosen the condition a little, to
catch URL prefixes with a stub.

Re "non-DOI URL stub," I mean that if the user (incorrectly) enters a full
URL like "http://my.blog.org/XXX-YYY" in the DOI field, it doesn't make
sense to confuse things further by transforming it to "
https://doi.org/XXX-YYY."

Re stripping other possible DOI prefixes from input values, I guess I'd ask
(1) whether we really want to do that, (2) whether we want to always do
that on the DOI field, regardless of processor settings (links are
currently an optional setting, which Zotero turns on), and (3) if we do
that, is it possible to reliably extract the DOI alone from a field with a
regexp?

Looking forward to the next iteration!

Frank


On Dec 10, 2017 13:30, "Rintze M. Zelle" <notifications@github.com> wrote:

So in short, I think the logic can be simplified to:

   1. Always reduce the DOI variable value to "10.4103/0976-500X.85940"
   format by deleting any kind of prefix.
   2. Include any domain in the DOI variable prefix in the hyperlink, but
   exclude any additional preceding text before the domain. E.g. with prefix="
   [https://doi.org/" the output should become [<a href="
   https://doi.org/10.4103/0976-500X.85940
   <https://doi.org/10.4103/0976-500X.85940>">https://doi.o
   rg/10.4103/0976-500X.85940</a>.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/Juris-M/citeproc-js/issues/57#issuecomment-350524547>,
or mute the thread

**Comment by fbennett** [comment]
Oops, correction. "catch URL prefixes with a stub" should have read "catch
URL prefixes preceded by other cruft."

**Comment by rmzelle** [comment]
> Re "non-DOI URL stub," I mean that if the user (incorrectly) enters a full URL like "http://my.blog.org/XXX-YYY" in the DOI field, it doesn't make sense to confuse things further by transforming it to " https://doi.org/XXX-YYY."

Do we really need to cover this case? I don't think we need to accommodate anybody abusing the DOI variable for non-DOI content, and it should be fine if the output is weird if you don't store DOIs in the field.

> Re stripping other possible DOI prefixes from input values, I guess I'd ask (1) whether we really want to do that, (2) whether we want to always do that on the DOI field, regardless of processor settings (links are currently an optional setting, which Zotero turns on), and (3) if we do that, is it possible to reliably extract the DOI alone from a field with a regexp?

(1) I think so. https://www.doi.org/doi_handbook/2_Numbering.html#2.2 is pretty clear that "doi:" or "http://doi.org" is not part of the DOI itself, and this can be made an explicit expectation in the CSL specification. Cleaning up DOI variable content might be more the responsibility of the reference manager (i.e. making sure that the CSL JSON is clean) than of the CSL processor, though.
(2) Idem.
(3) Assuming this doesn't need to work with shortDOIs (which according to https://www.crossref.org/display-guidelines/ aren't really DOIs), you could just trim up to the first "10.". The DOI format is a bit too amorphous too do much better than that, though, unfortunately (https://www.doi.org/doi_handbook/2_Numbering.html#2.2).

**Comment by fbennett** [comment]
Thanks for the links, sorry for being lazy. This is all pretty unsettling. I think I'll let the issue sit for a bit, and see what sort of complaints emerge.

Crossref seems not to be keen on shortDOIs, but they have enough traction that Chicago Manual [reportedly mentions them](https://github.com/zotero/zotero/issues/1371), so they're going to turn up in data.

**Comment by bwiernik** [comment]
> (3) Assuming this doesn't need to work with shortDOIs (which according to https://www.crossref.org/display-guidelines/ aren't really DOIs), you could just trim up to the first "10.". The DOI format is a bit too amorphous too do much better than that, though, unfortunately (https://www.doi.org/doi_handbook/2_Numbering.html#2.2).

I've seen enough users using shortDOIs to think the processor should probably handle them like full-length DOIs. The regex `/10\/[^\s]*[^\s\.,]/` does well for capturing both full DOIs and shortDOIs.

**Comment by rmzelle** [comment]
> The regex /10\/[^\s]*[^\s\.,]/ does well for capturing both full DOIs and shortDOIs.

The example on the Crossref page https://www.crossref.org/display-guidelines/#SHORTDOIs is https://doi.org/bwfc, which doesn't contain "10". (also shouldn't the `\/` be `\.`?)

**Comment by bwiernik** [comment]
(And non`\/` was correct.)

**Comment by bwiernik** [comment]
Okay, so for this item:
```[
	{
		"id": "http://zotero.org/users/1386342/items/FBZ76ZKB",
		"type": "article-journal",
		"title": "Development of four self-report measures of job stressors and strain: Interpersonal Conflict at Work Scale, Organizational Constraints Scale, Quantitative Workload Inventory, and Physical Symptoms Inventory",
		"container-title": "Journal of Occupational Health Psychology",
		"page": "356-367",
		"volume": "3",
		"issue": "4",
		"source": "APA PsycNET",
		"abstract": "Despite the widespread use of self-report measures of both job-related stressors and strains, relatively few carefully developed scales for which validity data exist are available. In this article, we discuss 3 job stressor scales (Interpersonal Conflict at Work Scale, Organizational Constraints Scale, and Quantitative Workload Inventory) and 1 job strain scale (Physical Symptoms Inventory). Using meta-analysis, we combined the results of 18 studies to provide estimates of relations between our scales and other variables. Data showed moderate convergent validity for the 3 job stressor scales, suggesting some objectivity to these self-reports. Norms for each scale are provided.",
		"URL": "http://doi.apa.org/getdoi.cfm?doi=10.1037/1076-8998.3.4.356",
		"DOI": "10/cng8cx",
		"ISSN": "1939-1307(Electronic);1076-8998(Print)",
		"shortTitle": "Development of four self-report measures of job stressors and strain",
		"journalAbbreviation": "JOHP",
		"language": "en",
		"author": [
			{
				"family": "Spector",
				"given": "Paul E."
			},
			{
				"family": "Jex",
				"given": "Steve M."
			}
		],
		"issued": {
			"date-parts": [
				[
					"1998"
				]
			]
		}
	}
]```

The full URL-form of the DOI is correctly wrapped in APA style and in Chicago (full-note), but only the DOI or shortDOI name is wrapped in American Chemical Society style (https://www.zotero.org/styles/american-chemical-society). I unfortunately don't see the difference causing this immediately at first glance. Just the (short or long) DOI name is wrapped in American Medical Association (which uses the doi: prefix).

With ACS, it should be:
(1) Spector, P. E.; Jex, S. M. Development of Four Self-Report Measures of Job Stressors and Strain: Interpersonal Conflict at Work Scale, Organizational Constraints Scale, Quantitative Workload Inventory, and Physical Symptoms Inventory. _JOHP_ **1998**, 3 (4), 356–367. [https://doi.org/10/cng8cx](https://doi.org/10/cng8cx).

But it renders as:
(1) Spector, P. E.; Jex, S. M. Development of Four Self-Report Measures of Job Stressors and Strain: Interpersonal Conflict at Work Scale, Organizational Constraints Scale, Quantitative Workload Inventory, and Physical Symptoms Inventory. _JOHP_ **1998**, 3 (4), 356–367. https://doi.org/[10/cng8cx](https://doi.org/10/cng8cx).

**Comment by fbennett** [comment]
Thanks - that sample is really helpful. It's not the form of the DOI that's tripping it up, but the prefix. In the ACS style, it's set as: `prefix=". https://doi.org/"` (with a leading period). Up to tag 1.1.211, the processor only recognizes a pure `https://doi.org/` prefix as an embedded protocol and domain, so it's treating the whole thing as a string prefix outside the anchor. I'll fix it up to separate the URL element from the string prefix in a subsequent release, coming up fairly soon.

**Comment by fbennett** [comment]
Fixed in tag release 1.1.212.
